File: Data Structures & Algorithms/search-for-word-ii/submission-1.py
class Node:
    def __init__(self):
        self.neigh = {}
        self.isEnd = False
        self.idx = -1

class Solution:
    def __init__(self):
        self.dirs = [(0,1), (0, -1), (1, 0), (-1, 0)]

    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        """
        1. build a trie tree for all the words
        2. search within the board to find the word on the trie tree
        """
        head = Node()
        idx = 0
        for w in words:
            cur = head
            for c in w:
                if c not in cur.neigh:
                    newNode = Node()
                    cur.neigh[c] = newNode
                cur = cur.neigh[c]
            cur.isEnd = True
            cur.idx = idx
            idx += 1
        ans = set()
        r = len(board)
        c = len(board[0])
        for i in range(r):
            for j in range(c):
                if board[i][j] in head.neigh:
                    self.backtrack(head, board, i, j, ans, words)       
        return list(ans)
    
    def backtrack(self, curr, board, i, j, ans, words):
        if board[i][j] == '#':
            return

        curr = curr.neigh[board[i][j]]
        tmp = board[i][j]
        board[i][j] = '#'
        if curr.idx >= 0:
            ans.add(words[curr.idx])
            curr.idx = -1
            # return

        for x, y in self.dirs:
            if i + x < 0 or i + x >= len(board) or j + y < 0 or j + y >= len(board[0]):
                continue
            r = i + x
            c = j + y
            if board[r][c] in curr.neigh:
                self.backtrack(curr, board, r, c, ans, words)
        
        board[i][j] = tmp

File: Data Structures & Algorithms/extra-characters-in-a-string/submission-1.cpp
struct Node {
  Node() { next = std::vector<std::unique_ptr<Node>>(26); }

  std::vector<std::unique_ptr<Node>> next;
  bool isEnd = false;
};

class Trie {
 public:
  Trie() { head = std::make_unique<Node>(); }

  ~Trie() {}

  void insert(const std::string& str) {
    auto cur = head.get();

    for (const char c : str) {
      if (!cur->next[c - 'a']) {
        cur->next[c - 'a'] = std::make_unique<Node>();
      }
      cur = cur->next[c - 'a'].get();
    }
    cur->isEnd = true;
  }

  bool search(const std::string& str) {
    auto cur = head.get();
    int idx = 0;
    for (const char c : str) {
      if (!cur->next[c - 'a']) {
        break;
      }
      cur = cur->next[c - 'a'].get();
      if (idx == str.size() - 1 && cur->isEnd) {
        return true;
      }
      ++idx;
    }
    return false;
  }

  std::unique_ptr<Node> head;
};

class Solution {
 public:
  int minExtraChar(string s, vector<string>& dictionary) {
    // 1. build a trie using dictonary
    Trie t; // issue with Trie t(); -> this is function declaration
    for (const auto& str : dictionary) {
      t.insert(str);
    }

    std::vector<int> dp(s.size(), -1);  // min extra chars to remove within s[i..end of str]
    return dfs(t, dp, 0, s);
  }

  int dfs(Trie& t, std::vector<int>& dp, int start, const string& s) {
    if (start == s.size()) {
      return 0;
    }

    if (dp[start] != -1) {
      return dp[start];
    }

    int ans = 1 + dfs(t, dp, start + 1, s);  // skip char at s[i]
    auto curr = t.head.get();

    for (int i = start; i < s.size(); ++i) {
      if (!curr->next[s[i] - 'a']) {
        break;
      }
      curr = curr->next[s[i] - 'a'].get();
      if (curr->isEnd) {
        ans = min(ans, dfs(t, dp, i + 1, s));
      }
    }

    dp[start] = ans;
    return ans;
  }
};